### DISCUSSION TITLE
Package manager discussion

### DISCUSSION THREAD

**Discussion by AnotherFoxGuy** [tenacityteam/tenacity-legacy#177] [General]
tenacity currently uses the [conan](https://conan.io/) package manager to manage some of the C++ dependencies

The question is, should tenacity keep using Conan to match the upstream builds or should it be switched to a different one? (Or completely none?)

Here is a list of possible alternatives:

* CMake meta-project: An cmake script that downloads and builds all the dependencies ([example](https://github.com/RigsOfRods/ror-dependencies), [example 2](https://github.com/OGRECave/ogre-next-deps))
* [vcpkg](https://github.com/microsoft/vcpkg) The C++ package manager by Microsoft (Note: since this is an Microsoft project, it has build in [telemetry](https://github.com/microsoft/vcpkg/blob/master/docs/about/privacy.md))
* [hunter](https://github.com/cpp-pm/hunter) A CMake script to manage dependencies 
* [CPM](https://github.com/cpm-cmake/CPM.cmake) An CMake script that uses the build-in FetchContent CMake module

**Comment by Songtech-0912** [comment]
I believe that keeping things to `conan` for now is the best choice, as changing the build system is going to make things more difficult for us and divert attention from these things - things that are likely more important:

- Getting project governance cleared up
- Building our identity with a new logo, website, and community
- Spreading awareness of Tenacity and dispelling notions that it's just "an unofficial version of Audacity"
- Making a roadmap for stuff to do next

The alternatives you posted are all very good, but let's keep them on the shelf until sometime when it's ready to consider this issue.

**Reply by Be-ing** [reply]
Considering that Windows builds take 30 minutes and I don't know how to improve this with Conan (#176), I think it is worth discussing now.

**Reply by falkTX** [reply]
well I just found out we are targeting the wrong branch anyway. a lot of work is being done on release-3.0.3 branch
some conan related changes there too.

https://github.com/audacity/audacity/compare/release-3.0.3

**Reply by Be-ing** [reply]
I don't see anything in that branch that I think would affect the issue I stumbled on in #176.

**Comment by Be-ing** [comment]
> CMake meta-project: An cmake script that downloads and builds all the dependencies (example, example 2)

This would be an unmaintainable mess. The point of a package manager is that updating dependencies is easy.

> vcpkg The C++ package manager by Microsoft (Note: since this is an Microsoft project, it has build in telemetry)

The telemetry can be disabled easily, so I don't think this is really a problem.

> hunter A CMake script to manage dependencies

The package list is really small.

> CPM An CMake script that uses the build-in FetchContent CMake module

[Limitations](https://github.com/cpm-cmake/CPM.cmake#limitations) are significant.

There is MSYS2, but that only helps on Windows, not macOS. AFAIK vcpkg seems to be the only viable alternative to Conan. I would be happy to be wrong about that.

**Comment by ErikPrantare** [comment]
Is there any issue with conan that is solved by the other ones?

**Reply by Be-ing** [reply]
Yes, compiler caching: #176

**Reply by Be-ing** [reply]
The issue with the Ninja generator with MSVC has been fixed in #184. However, I find Conan quite an ugly solution with how it requires the CMakeLists.txt files to explicitly support it. vcpkg is nicer because no special hacks are required to support it with CMake; all it requires is passing a single variable to CMake:

```
cmake -B [build directory] -S . -DCMAKE_TOOLCHAIN_FILE=[path to vcpkg]/scripts/buildsystems/vcpkg.cmake
```

Also Conan is written in Python and requiring Python to build C++ is ugly.

**Reply by falkTX** [reply]
upstream uses conan, so for the sake of maintainability let's keep that.

also, please stop pushing for vcpkg, do you work for them or something? If I didn't know best, I would say they pay you lol.
It is getting annoying your constant push for it at this point.
thanks for understanding.

**Reply by Be-ing** [reply]
I am only presenting it as an option because I have experience with it, know it works well, and is easy to maintain. Also for me personally, Tenacy and Mixxx share a lot of dependencies, so if they both used vcpkg it would be easy for me to maintain both. I am not saying we have to switch to it, but I think it is worth considering the available options.

**Reply by Be-ing** [reply]
> I find Conan quite an ugly solution with how it requires the CMakeLists.txt files to explicitly support it.

This is turning out to be quite a mess and flaky when changing options and rebuilding. The more I attempt to work with the tangled hacks using Conan and the vendored dependencies, the less I think it is a good idea to try to maintain them. Fortunately they are well isolated inside the cmake-proxies directory. #228 is a proof-of-concept how the hacks can be bypassed for normal CMake behavior using `find_package` without needing to touch Audacity's hacks in cmake-proxies, just some small modifications to the root CMakeLists.txt and src/CMakeLists.txt to use `find_package` instead of
```
add_subdirectory( "cmake-proxies" )
resolve_conan_dependencies()
```
`find_package` would work easily with vcpkg. The CMake modules copied from Mixxx are already known to work with Linux distro package managers as well as vcpkg without CMakeLists.txt needing any conditional expressions depending on whether vcpkg is used or not.

**Reply by Be-ing** [reply]
Audacity seems [intent](https://github.com/audacity/audacity/issues/1187) on doubling down on their tangled hacks and telling Linux distros to use Conan instead of their own packages, so I think bypassing that mess as #228 does would be more maintainable than trying to fix it.

**Comment by falkTX** [comment]
conan is used upstream, IMO it makes perfect sense to keep it as long as tenacity is still running in sync with it.

**Comment by nbsp** [comment]
Conan doesn't support musl libc natively, which means Alpine Linux users might encounter problems building from source.
I've actually been affected by this when trying to write an Alpine build manifest for builds.sr.ht.

The Alpine maintainer for Audacity stated they'll only package the project if we support musl upstream, so we can't use a package manager that doesn't.

**Reply by Be-ing** [reply]
> if we support musl upstream

What does this mean? Running CI builds with musl libc? I don't think that's really necessary. Once we get Tenacity building with system packages instead of requiring vendored forks, what package manager we use for official builds should be irrelevant for people packaging Tenacity for other package managers.

**Reply by nbsp** [reply]
Well, running CI builds on Alpine is easy with builds.sr.ht. That's not the problem.

"should be irrelevant" is all well and good, but musl/glibc inconsistencies can break some stuff, and we need to make sure it doesn't happen.

**Reply by Be-ing** [reply]
If all of Tenacity's dependencies are packaged in Alpine, running CI builds on Alpine should be simple and the package manager we use for Windows and macOS should not matter.

**Comment by purplesyringa** [comment]
If Conan doesn't support musl and other package managers don't support this or that dependency, is using submodules for unsupported dependencies so much of a problem?

**Reply by Be-ing** [reply]
I don't understand how Git submodules would help anything.

**Reply by purplesyringa** [reply]
Git submodules solve the same problem as package managers: you don't have to copy code from another repository, you can use the project's CMakeLists instead of using handcrafted cmake proxies like Audacity does at the moment (provided that the dependency supports cmake of course), updating to a newer version is easy and you immediately know what patches were applied (if any) to the dependencies instead of reading (quite outdated) audacity-patches.txt. What else do you need from a package manager?

**Reply by Be-ing** [reply]
Both Conan and vcpkg can easily package prebuilt binaries for dependencies so developers don't need to build the entire dependency graph on their own computer. Perhaps for Tenacity this isn't such a big deal because it doesn't take all that long to build its dependency graph. If ccache or sccache is used when the dependencies are built, that might not be too much of a barrier for new contributors.

**Reply by falkTX** [reply]
on Linux we do not want to use conan but always use system libraries.
so whatever conan supports (or not) is irrelevant there.

**Reply by purplesyringa** [reply]
Okay... Please keep in mind that when using local libraries Audacity links them statically so this may be somewhat less efficient.

**Reply by Be-ing** [reply]
Using Conan or vcpkg for a Flatpak and/or AppImage build could be useful so dependencies for every OS are managed in one place.

**Reply by falkTX** [reply]
> Okay... Please keep in mind that when using local libraries Audacity links them statically so this may be somewhat less efficient.

except this isn't true. it will link to shared or static, whatever is available.
If only shared version is available, that will be used.

**Reply by purplesyringa** [reply]
I might be mistaken but I think cmake configuration suggests that when Audacity builds the library manually it links it statically?

**Reply by falkTX** [reply]
it might have a preference for static libs, I dont know it myself.
but from a compiler's perspective, you can pick if you prefer one or the other, but cant ask to link againt only 1 variant. only way around it for compilation would be to find and specify the static lib file instead of the usual "-L path -l lib"

**Reply by purplesyringa** [reply]
I think it is also kind of obvious that static libs are used for local build, because the .so's are not installed next to audacity binary.

**Reply by Be-ing** [reply]
If we used Git submodules it would be up to us to keep them up to date. With a package manager, other people are helping maintain Tenacity's dependencies too.

**Reply by purplesyringa** [reply]
Why? I mean... how is updating a git submodule to a newer commit different from updating conan packages to a new version?

**Comment by Be-ing** [comment]
@falkTX you have mentioned cross compiling from Linux to Windows and macOS. What options are available for package managers for cross compiled dependencies? I think Conan and vcpkg can both be used for this.

**Reply by falkTX** [reply]
I think you know the same as me. msys2 can be a candidate for mingw.
macOS has homebrew and macports.

I know most dont agree, but since this is an opensource project, building stuff ourselves statically is also an option.
So the final binary doesnt have a bunch of dependency dlls, but is a single executable.
But not something to do now anyway.
